Jenkinsfile

```diff
@@ -0,0 +1,306 @@
+        PATH = "${env.JAVA_HOME}/bin:${env.PATH}"
+      }
+      
+      when { 
+        beforeAgent true
+        environment name: 'linux_ci', value: 'true' 
+      }
+      steps {
+        withEnvWrapper() {
+          unstash 'source'
+          dir("${BASE_DIR}"){    
+            sh """#!/bin/bash
+            ./mvnw clean package -DskipTests=true  -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
+            """
+          }
+        }
+      }
+    }
+    stage('Parallel stages') {
+      failFast true
+      parallel {
+        stage('test') {
+          agent { label 'linux && immutable' }
+          environment {
+            HOME = "${env.HUDSON_HOME}"
+            JAVA_HOME = "${env.HOME}/.java/java10"
+            PATH = "${env.JAVA_HOME}/bin:${env.PATH}"
+          }
+          
+          when { 
+            beforeAgent true
+            environment name: 'test_ci', value: 'true' 
+          }
+          steps {
+            withEnvWrapper() {
+              unstash 'source'
+              dir("${BASE_DIR}"){    
+                sh """#!/bin/bash
+                ./mvnw clean verify --batch-mode  -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
+                """
+                codecov('apm-agent-java')
+              }
+            }
+          }
+          post { 
+            always {
+              coverageReport("${BASE_DIR}/build/coverage")
+              junit(allowEmptyResults: true, 
+                keepLongStdio: true, 
+                testResults: "${BASE_DIR}/**/junit-*.xml")
+            }
+          }
+        }
+        stage('Benchmarks') {
+          agent { label 'metal' }
+          environment {
+            HOME = "${env.HUDSON_HOME}"
+            JAVA_HOME = "${env.HOME}/.java/java10"
+            PATH = "${env.JAVA_HOME}/bin:${env.PATH}"
+          }
+          
+          when { 
+            beforeAgent true
+            allOf { 
+              //branch 'master';
+              environment name: 'bench_ci', value: 'true' 
+            }
+          }
+          steps {
+            withEnvWrapper() {
+              unstash 'source'
+              dir("${BASE_DIR}"){
+                script {
+                  env.COMMIT_ISO_8601 = sh(script: 'git log -1 -s --format=%cI', returnStdout: true).trim()
+                  env.NOW_ISO_8601 = sh(script: 'date -u "+%Y-%m-%dT%H%M%SZ"', returnStdout: true).trim()
+                  env.RESULT_FILE = "apm-agent-benchmark-results-${env.COMMIT_ISO_8601}.json"
+                  env.BULK_UPLOAD_FILE = "apm-agent-bulk-${env.NOW_ISO_8601}.json"
+                }
+                sh """#!/bin/bash
+                ./scripts/jenkins/run-benchmarks.sh
+                """
+              }
+            }
+          } 
+          post {
+            always {
+              sendBenchmarks(file: "${BASE_DIR}/${BULK_UPLOAD_FILE}", index: "benchmark-java", bulk: true)
```

Comment:
where is sendBenchmarks defined?

Reply:
The step is defined in the pipeline library, I am testing the step for the Java agent, probably I have to make some adjustments.